File: 2018/gabriel_python3/src/day5/5.2.py
#!/usr/bin/env python3

import re
import sys
sys.setrecursionlimit(5000)  # hint that this is not such a good idea ...

def match (x, y):
    return x.lower() == y.lower() and x != y

def final_reduce(part1, part2):
    if not part1:
        return part2
    elif not part2:
        return part1
    elif match(part1[-1], part2[0]):
        return final_reduce(part1[:-1], part2[1:])
    else:
        return part1 + part2

def reduce(seq):
    l = len(seq)
    if l < 2:
        return seq
    elif l == 2:
        if match(seq[0], seq[1]):
            return ''
        else:
            return seq
    else:
        part1 = seq[:l//2]
        part2 = seq[l//2:]
        rv1 = reduce(part1)
        rv2 = reduce(part2)
        if part1 == rv1 and part2 == rv2:
            return final_reduce(part1, part2)
        else:
            return reduce(rv1 + rv2)


# 10901 too low
filename = 'input'

with open(filename) as f:
    step1 = reduce(f.read().strip())
    minlen = len(step1)
    for letter in 'azertyuiopsdfghjklmwxcvbn':
        tmp = re.sub('[' + letter + letter.upper() + ']', '', step1)
        tmp = reduce(tmp)
        if len(tmp) < minlen:
            minlen = len(tmp)
            print('removed:', letter)

    print(minlen)

File: 2018/gabriel_rust/src/day5/5-2.rs
use std::fs::File;
use std::io::prelude::*;

// match is a reserved keyword
fn charmatch(x: char, y: char) -> bool {
    let lx = x.to_lowercase().last();
    let ly = y.to_lowercase().last();
    return lx == ly && x != y;
}

fn final_reduce(mut part1: String, mut part2: String) -> String {
    if part1.is_empty() {
        return part2;
    } else if part2.is_empty() {
        return part1;
    } else if charmatch(part1.chars().last().unwrap(), part2.chars().nth(0).unwrap()) {
        part1.pop();
        part2.remove(0);
        return final_reduce(part1, part2);
    } else {
        let mut rv = String::new();
        rv.push_str(&part1);
        rv.push_str(&part2);
        return rv;
    }
}
    

fn reduce(seq: String) -> String {
    let l = seq.len();
    if l < 2 {
        return seq;
    } else if l == 2 {
        if charmatch(seq.chars().nth(0).unwrap(), seq.chars().last().unwrap()) {
            return String::new();
        } else {
            return seq;
        }
    } else {
        let (part1, part2) = seq.split_at(l/2);
        let mut rv1 = reduce(part1.to_string());
        let rv2 = reduce(part2.to_string());
        if part1 == rv1 && part2 == rv2 {
            return final_reduce(rv1, rv2);
        } else {
            rv1.push_str(&rv2);
            return reduce(rv1);
        }
    }
}

fn main() {
    let filename = "input";

    let mut f = File::open(filename).expect("file not found");

    let mut content = String::new();
    f.read_to_string(&mut content)
        .expect("something went wrong reading the file");

    content.pop();  // remove newline char

    let mut minlen = content.len();
    let step1 = reduce(content);
    for letter in "azertyuiopsdfghjklmwxcvbn".chars() {
        let mut tmp: String = step1.replace(letter, "");
        let upper = letter.to_uppercase().last().unwrap();
        tmp = tmp.replace(upper, "");
        tmp = reduce(tmp);
        if tmp.len() < minlen {
            minlen = tmp.len();
            println!("removed: {}", letter);
        }
    }


    println!("{}", minlen);
}
